File: operators/src/main/java/net/bertag/operators/Ops.java
package net.bertag.operators;

import net.bertag.operators.api.Addable;
import net.bertag.operators.api.Dividable;
import net.bertag.operators.api.Multipliable;
import net.bertag.operators.api.Scalable;
import net.bertag.operators.api.Subtractable;

import java.util.function.BiFunction;
import java.util.function.BinaryOperator;
import java.util.function.Function;

/**
 * Utility class defining null-safe transformation operations.
 */
public class Ops {

    /**
     * Adds the two numbers together (`a + b`) if at least one input is non-null (substituting 0 for the other input if
     * it is null), or returning null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the sum as described, or null if both inputs are null
     */
    public static Integer plus(Integer a, Integer b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0) + defaultIfNull(b, 0);
    }

    /**
     * Adds the two numbers together ({@code a + b}) if at least one input is non-null (substituting 0 for the other
     * input if it is null), or returning null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the sum as described, or null if both inputs are null
     */
    public static Long plus(Long a, Long b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0L) + defaultIfNull(b, 0L);
    }

    /**
     * Adds the two numbers together ({@code a + b}) if at least one input is non-null (substituting 0 for the other
     * input if it is null), or returning null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the sum as described, or null if both inputs are null
     */
    public static Short plus(Short a, Short b) {
        return allNull(a, b) ? null : (short) (defaultIfNull(a, (short) 0) + defaultIfNull(b, (short) 0));
    }

    /**
     * Adds the two numbers together ({@code a + b}) if at least one input is non-null (substituting 0 for the other
     * input if it is null), or returning null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the sum as described, or null if both inputs are null
     */
    public static Float plus(Float a, Float b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0f) + defaultIfNull(b, 0f);
    }

    /**
     * Adds the two numbers together ({@code a + b}) if at least one input is non-null (substituting 0 for the other
     * input if it is null), or returning null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the sum as described, or null if both inputs are null
     */
    public static Double plus(Double a, Double b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0d) + defaultIfNull(b, 0d);
    }

    /**
     * Adds the two objects together ({@code a + b}) using {@link Addable#plus(Object)} if both inputs are non-null; if
     * either input is null, the other input is returned.
     *
     * @param a some nullable object
     * @param b another nullable object
     * @return the sum as described, or null if both inputs are null
     * @param <T> the type of objects being added
     */
    public static <T extends Addable<T>> T plus(T a, T b) {
        return op(a, b, Addable::plus);
    }

    /**
     * Subtracts the second number from the first ({@code a - b}) if at least one input is non-null (substituting 0 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the difference as described, or null if both inputs are null
     */
    public static Integer minus(Integer a, Integer b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0) - defaultIfNull(b, 0);
    }

    /**
     * Subtracts the second number from the first ({@code a - b}) if at least one input is non-null (substituting 0 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the difference as described, or null if both inputs are null
     */
    public static Long minus(Long a, Long b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0L) - defaultIfNull(b, 0L);
    }

    /**
     * Subtracts the second number from the first ({@code a - b}) if at least one input is non-null (substituting 0 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the difference as described, or null if both inputs are null
     */
    public static Short minus(Short a, Short b) {
        return allNull(a, b) ? null : (short) (defaultIfNull(a, (short) 0) - defaultIfNull(b, (short) 0));
    }

    /**
     * Subtracts the second number from the first ({@code a - b}) if at least one input is non-null (substituting 0 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the difference as described, or null if both inputs are null
     */
    public static Float minus(Float a, Float b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0f) - defaultIfNull(b, 0f);
    }

    /**
     * Subtracts the second number from the first ({@code a - b}) if at least one input is non-null (substituting 0 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the difference as described, or null if both inputs are null
     */
    public static Double minus(Double a, Double b) {
        return allNull(a, b) ? null : defaultIfNull(a, 0d) - defaultIfNull(b, 0d);
    }

    /**
     * Subtracts the second object from the first ({@code a - b}) using {@link Subtractable#minus(Object)} if both
     * inputs are non-null; if either input is null, the other input is returned.
     *
     * @param a some nullable object
     * @param b another nullable object
     * @return the difference as described, or null if both inputs are null
     * @param <T> the type of objects being subtracted
     */
    public static <T extends Subtractable<T>> T minus(T a, T b) {
        return op(a, b, Subtractable::minus);
    }

    /**
     * Multiplies the two numbers together ({@code a * b}) if at least one input is non-null (substituting 1 for the
     * other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the product as described, or null if both inputs are null
     */
    public static Integer times(Integer a, Integer b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1) * defaultIfNull(b, 1);
    }

    /**
     * Multiplies the two numbers together ({@code a * b}) if at least one input is non-null (substituting 1 for the
     * other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the product as described, or null if both inputs are null
     */
    public static Long times(Long a, Long b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1L) * defaultIfNull(b, 1L);
    }

    /**
     * Multiplies the two numbers together ({@code a * b}) if at least one input is non-null (substituting 1 for the
     * other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the product as described, or null if both inputs are null
     */
    public static Short times(Short a, Short b) {
        return allNull(a, b) ? null : (short) (defaultIfNull(a, (short) 1) * defaultIfNull(b, (short) 1));
    }

    /**
     * Multiplies the two numbers together ({@code a * b}) if at least one input is non-null (substituting 1 for the
     * other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the product as described, or null if both inputs are null
     */
    public static Float times(Float a, Float b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1f) * defaultIfNull(b, 1f);
    }

    /**
     * Multiplies the two numbers together ({@code a * b}) if at least one input is non-null (substituting 1 for the
     * other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the product as described, or null if both inputs are null
     */
    public static Double times(Double a, Double b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1d) * defaultIfNull(b, 1d);
    }

    /**
     * Multiplies the two objects together ({@code a * b}) using {@link Multipliable#times(Object)} if both inputs are
     * non-null; if either input is null, the other input is returned.
     *
     * @param a some nullable object
     * @param b another nullable object
     * @return the product as described, or null if both inputs are null
     * @param <T> the type of objects being multiplied
     */
    public static <T extends Multipliable<T>> T times(T a, T b) {
        return op(a, b, Multipliable::times);
    }

    /**
     * Divides the second number from the first ({@code a / b}) if at least one input is non-null (substituting 1 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the quotient as described, or null if both inputs are null
     */
    public static Integer div(Integer a, Integer b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1) / defaultIfNull(b, 1);
    }

    /**
     * Divides the second number from the first ({@code a / b}) if at least one input is non-null (substituting 1 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the quotient as described, or null if both inputs are null
     */
    public static Long div(Long a, Long b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1L) / defaultIfNull(b, 1L);
    }

    /**
     * Divides the second number from the first ({@code a / b}) if at least one input is non-null (substituting 1 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the quotient as described, or null if both inputs are null
     */
    public static Short div(Short a, Short b) {
        return allNull(a, b) ? null : (short) (defaultIfNull(a, (short) 1) / defaultIfNull(b, (short) 1));
    }

    /**
     * Divides the second number from the first ({@code a / b}) if at least one input is non-null (substituting 1 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the quotient as described, or null if both inputs are null
     */
    public static Float div(Float a, Float b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1f) / defaultIfNull(b, 1f);
    }

    /**
     * Divides the second number from the first ({@code a / b}) if at least one input is non-null (substituting 1 for
     * the other input if it is null), or null if both inputs are null.
     *
     * @param a some nullable number
     * @param b another nullable number
     * @return the quotient as described, or null if both inputs are null
     */
    public static Double div(Double a, Double b) {
        return allNull(a, b) ? null : defaultIfNull(a, 1d) / defaultIfNull(b, 1d);
    }

    /**
     * Divides the second object from the first ({@code a / b}) using {@link Dividable#div(Object)} if both inputs are
     * non-null; if either input is null, the other input is returned.
     *
     * @param a some nullable object
     * @param b another nullable object
     * @return the quotient as described, or null if both inputs are null
     * @param <T> the type of objects being divided
     */
    public static <T extends Dividable<T>> T div(T a, T b) {
        return op(a, b, Dividable::div);
    }

    /**
     * Scales a number by a given factor if both are non-null.  If the {@code value} is null, null is returned; if the
     * {@code factor} is null, no scaling will be applied.
     *
     * @param value some nullable number
     * @param factor a nullable factor
     * @return the scaled result rounded to the nearest integer as described, or null if {@code value} is null
     */
    public static Integer scale(Integer value, Double factor) {
        return value != null ? (int) Math.round(value * defaultIfNull(factor, 1d)) : null;
    }

    /**
     * Scales a number by a given factor if both are non-null.  If the {@code value} is null, null is returned; if the
     * {@code factor} is null, no scaling will be applied.
     *
     * @param value some nullable number
     * @param factor a nullable factor
     * @return the scaled result rounded to the nearest (long) integer as described, or null if {@code value} is null
     */
    public static Long scale(Long value, Double factor) {
        return value != null ? Math.round(value * defaultIfNull(factor, 1d)) : null;
    }

    /**
     * Scales a number by a given factor if both are non-null.  If the {@code value} is null, null is returned; if the
     * {@code factor} is null, no scaling will be applied.
     *
     * @param value some nullable number
     * @param factor a nullable factor
     * @return the scaled result rounded to the nearest (short) integer as described, or null if {@code value} is null
     */
    public static Short scale(Short value, Double factor) {
        return value != null ? (short) Math.round(value * defaultIfNull(factor, 1d)) : null;
    }

    /**
     * Scales a number by a given factor if both are non-null.  If the {@code value} is null, null is returned; if the
     * {@code factor} is null, no scaling will be applied.
     *
     * @param value some nullable number
     * @param factor a nullable factor
     * @return the scaled result as described, or null if {@code value} is null
     */
    public static Float scale(Float value, Double factor) {
        return value != null ? (float) (value * defaultIfNull(factor, 1d)) : null;
    }

    /**
     * Scales a number by a given factor if both are non-null.  If the {@code value} is null, null is returned; if the
     * {@code factor} is null, no scaling will be applied.
     *
     * @param value some nullable number
     * @param factor a nullable factor
     * @return the scaled result as described, or null if {@code value} is null
     */
    public static Double scale(Double value, Double factor) {
        return value != null ? value * defaultIfNull(factor, 1d) : null;
    }

    /**
     * Scales a {@link Scalable} object by a given factor if both are non-null.  If the first argument is null, null is
     * returned; if the factor is null, no scaling will be applied.
     *
     * @param value some nullable number
     * @param factor a nullable factor
     * @return the scaled result as described, or null if {@code value} is null
     */
    public static <T extends Scalable<T>> T scale(T value, Double factor) {
        return apply(value, factor, Scalable::scale);
    }

    /**
     * Applies the given transformation to an object if it is not null.  If the value is null, then null is returned.
     *
     * @param value some nullable value
     * @param opFunction the operation to perform
     * @return the result as described, or null if {@code value} is null
     * @param <T> the type of object being operated on
     * @param <R> the type of result
     */
    public static <T, R> R map(T value, Function<T, R> opFunction) {
        return value != null ? opFunction.apply(value) : null;
    }

    /**
     * Performs the given operation on the given inputs if they are both non-null.  If one input is null, the other
     * input is returned; if both are null, null is returned.
     *
     * @param a some nullable object
     * @param b another nullable object
     * @param opFunction the operation to perform
     * @return the result as described, or null if both inputs are null
     * @param <T> the type of objects being operated upon
     */
    public static <T> T op(T a, T b, BinaryOperator<T> opFunction) {
        if (a == null) return b;
        else if (b == null) return a;
        else return opFunction.apply(a, b);
    }

    /**
     * Applies the given transformation to an object if the other input is non-null.  If the receiver (Kotlin) — or first
     * argument (Java) — is null, then null is returned; if `other` is null, then the receiver will be returned.
     *
     * @param receiver some nullable object
     * @param other another nullable object
     * @param opFunction the operation to perform
     * @return the result as described, or null if {@code receiver} is null
     * @param <T> the type of objects being operated upon
     */
    public static <T, U> T apply(T receiver, U other, BiFunction<T, U, T> opFunction) {
        if (receiver == null) return null;
        else if (other == null) return receiver;
        else return opFunction.apply(receiver, other);
    }

    /**
     * Determines if both objects are null.
     *
     * @param a some nullable object
     * @param b another nullable object
     * @return true if both objects are null; false if either or both is non-null
     */
    private static boolean allNull(Object a, Object b) {
        return a == null && b == null;
    }

    // Taken from https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/src-html/org/apache/commons/lang3/ObjectUtils.html.
    // Original code licensed as follows:
    //
    // Licensed to the Apache Software Foundation (ASF) under one or more
    // contributor license agreements.  See the NOTICE file distributed with
    // this work for additional information regarding copyright ownership.
    // The ASF licenses this file to You under the Apache License, Version 2.0
    // (the "License"); you may not use this file except in compliance with
    // the License.  You may obtain a copy of the License at
    //
    //      http://www.apache.org/licenses/LICENSE-2.0
    //
    // Unless required by applicable law or agreed to in writing, software
    // distributed under the License is distributed on an "AS IS" BASIS,
    // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    // See the License for the specific language governing permissions and
    // limitations under the License.
    /**
     * <p>Returns a default value if the object passed is {@code null}.</p>
     *
     * <pre>
     * ObjectUtils.defaultIfNull(null, null)      = null
     * ObjectUtils.defaultIfNull(null, "")        = ""
     * ObjectUtils.defaultIfNull(null, "zz")      = "zz"
     * ObjectUtils.defaultIfNull("abc", *)        = "abc"
     * ObjectUtils.defaultIfNull(Boolean.TRUE, *) = Boolean.TRUE
     * </pre>
     *
     * @param <T> the type of the object
     * @param object  the {@code Object} to test, may be {@code null}
     * @param defaultValue  the default value to return, may be {@code null}
     * @return {@code object} if it is not {@code null}, defaultValue otherwise
     */
    private static <T> T defaultIfNull(T object, T defaultValue) {
        return object != null ? object : defaultValue;
    }
}

File: src/main/kotlin/net/bertag/operators/Ops.kt
@file:JvmName("Ops")
package net.bertag.operators

import net.bertag.operators.api.Addable
import net.bertag.operators.api.Dividable
import net.bertag.operators.api.Multipliable
import net.bertag.operators.api.Scalable
import net.bertag.operators.api.Subtractable
import java.util.function.BiFunction
import java.util.function.BinaryOperator
import kotlin.math.roundToInt
import kotlin.math.roundToLong

/**
 * Adds the two numbers together (e.g.: `a + b`) if at least one input is non-null (substituting 0 for the
 * other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the sum as described
 */
operator fun Int?.plus(other: Int?) = if (allNull(this, other)) null else (this ?: 0) + (other ?: 0)

/**
 * Adds the two numbers together (e.g.: `a + b`) if at least one input is non-null (substituting 0 for the
 * other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the sum as described
 */
operator fun Long?.plus(other: Long?) = if (allNull(this, other)) null else (this ?: 0L) + (other ?: 0L)

/**
 * Adds the two numbers together (e.g.: `a + b`) if at least one input is non-null (substituting 0 for the
 * other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the sum as described
 */
operator fun Short?.plus(other: Short?) = if (allNull(this, other)) null else ((this ?: 0) + (other ?: 0)).toShort()

/**
 * Adds the two numbers together (e.g.: `a + b`) if at least one input is non-null (substituting 0 for the
 * other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the sum as described
 */
operator fun Float?.plus(other: Float?) = if (allNull(this, other)) null else (this ?: 0F) + (other ?: 0F)

/**
 * Adds the two numbers together (e.g.: `a + b`) if at least one input is non-null (substituting 0 for the
 * other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the sum as described
 */
operator fun Double?.plus(other: Double?) = if (allNull(this, other)) null else (this ?: 0.0) + (other ?: 0.0)

/**
 * Adds the two objects together (e.g.: `a * b`) using [Addable.plus] if both inputs are non-null.
 * If one input is null, the other input is returned; if both are null, null is returned.
 *
 * @receiver some nullable object
 * @param other another nullable object
 * @return the sum as described
 * @param <T> the type of objects being subtracted
 */
operator fun <T : Addable<T>> T?.plus(other: T?) = this.op(other) { x, y -> x + y }

/**
 * Subtracts the second number from the first (e.g.: `a - b`) if at least one input is non-null (substituting
 * 0 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Int?.minus(other: Int?) = if (allNull(this, other)) null else (this ?: 0) - (other ?: 0)

/**
 * Subtracts the second number from the first (e.g.: `a - b`) if at least one input is non-null (substituting
 * 0 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Long?.minus(other: Long?) = if (allNull(this, other)) null else (this ?: 0L) - (other ?: 0L)

/**
 * Subtracts the second number from the first (e.g.: `a - b`) if at least one input is non-null (substituting
 * 0 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Short?.minus(other: Short?) = if (allNull(this, other)) null else ((this ?: 0) - (other ?: 0)).toShort()

/**
 * Subtracts the second number from the first (e.g.: `a - b`) if at least one input is non-null (substituting
 * 0 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Float?.minus(other: Float?) = if (allNull(this, other)) null else (this ?: 0F) - (other ?: 0F)

/**
 * Subtracts the second number from the first (e.g.: `a - b`) if at least one input is non-null (substituting
 * 0 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Double?.minus(other: Double?) = if (allNull(this, other)) null else (this ?: 0.0) - (other ?: 0.0)

/**
 * Subtracts the second object from the first (e.g.: `a - b`) using [Subtractable.minus] if both
 * inputs are non-null.  If one input is null, the other input is returned; if both are null, null is returned.
 *
 * @receiver some nullable object
 * @param other another nullable object
 * @return the difference as described
 * @param <T> the type of objects being subtracted
</T> */
operator fun <T : Subtractable<T>> T?.minus(other: T?) = this.op(other) { x, y -> x - y }

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the product as described
 */
operator fun Int?.times(other: Int?) = if (allNull(this, other)) null else (this ?: 1) * (other ?: 1)

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the product as described
 */
operator fun Long?.times(other: Long?) = if (allNull(this, other)) null else (this ?: 1L) * (other ?: 1L)

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the product as described
 */
operator fun Short?.times(other: Short?) = if (allNull(this, other)) null else ((this ?: 1) * (other ?: 1)).toShort()

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the product as described
 */
operator fun Float?.times(other: Float?) = if (allNull(this, other)) null else (this ?: 1F) * (other ?: 1F)

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the product as described
 */
operator fun Double?.times(other: Double?) = if (allNull(this, other)) null else (this ?: 1.0) * (other ?: 1.0)

/**
 * Multiplies the two objects together (e.g.: `a * b`) using [Multipliable.times] if both inputs are
 * non-null.  If one input is null, the other input is returned; if both are null, null is returned.
 *
 * @receiver some nullable object
 * @param other another nullable object
 * @return the product as described
 * @param <T> the type of objects being multiplied
 */
operator fun <T : Multipliable<T>> T?.times(other: T?) = this.op(other) { x, y -> x * y }

/**
 * Divides the second number from the first (e.g.: `a / b`) if at least one input is non-null (substituting
 * 1 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Int?.div(other: Int?) = if (allNull(this, other)) null else (this ?: 1) / (other ?: 1)

/**
 * Divides the second number from the first (e.g.: `a / b`) if at least one input is non-null (substituting
 * 1 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Long?.div(other: Long?) = if (allNull(this, other)) null else (this ?: 1L) / (other ?: 1L)

/**
 * Divides the second number from the first (e.g.: `a / b`) if at least one input is non-null (substituting
 * 1 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Short?.div(other: Short?) = if (allNull(this, other)) null else ((this ?: 1) / (other ?: 1)).toShort()

/**
 * Divides the second number from the first (e.g.: `a / b`) if at least one input is non-null (substituting
 * 1 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Float?.div(other: Float?) = if (allNull(this, other)) null else (this ?: 1F) / (other ?: 1F)

/**
 * Divides the second number from the first (e.g.: `a / b`) if at least one input is non-null (substituting
 * 1 for the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param other another nullable number
 * @return the difference as described
 */
operator fun Double?.div(other: Double?) = if (allNull(this, other)) null else (this ?: 1.0) / (other ?: 1.0)

/**
 * Divides the second object from the first (e.g.: `a / b`) using [Dividable.div] if both inputs
 * are non-null.  If one input is null, the other input is returned; if both are null, null is returned.
 *
 * @receiver some nullable object
 * @param other another nullable object
 * @return the product as described
 * @param <T> the type of objects being divided
 */
operator fun <T : Dividable<T>> T?.div(other: T?) = op(other) { x, y -> x / y }

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param factor another nullable number
 * @return the product as described
 */
fun Int?.scale(factor: Double?) = if (this == null) null else (this * (factor ?: 1.0)).roundToInt()

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param factor another nullable number
 * @return the product as described
 */
fun Long?.scale(factor: Double?) = if (this == null) null else (this * (factor ?: 1.0)).roundToLong()

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param factor another nullable number
 * @return the product as described
 */
fun Short?.scale(factor: Double?) = if (this == null) null else (this * (factor ?: 1.0)).roundToInt().toShort()

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param factor another nullable number
 * @return the product as described
 */
fun Float?.scale(factor: Double?) = if (this == null) null else (this * (factor ?: 1.0)).toFloat()

/**
 * Multiplies the two numbers together (e.g.: `a * b`) if at least one input is non-null (substituting 1 for
 * the other input if it is null), or null if both inputs are null.
 *
 * @receiver some nullable number
 * @param factor another nullable number
 * @return the product as described
 */
fun Double?.scale(factor: Double?) = if (this == null) null else (this * (factor ?: 1.0))

/**
 * Scales this object using [Scalable.scale] if both inputs are non-null.  If `this` object is null, null is returned;
 * if the factor is null, then the other input is returned as is.
 *
 * @receiver some nullable object
 * @param factor another nullable object
 * @return the product as described
 * @param <T> the type of objects being multiplied
 */
fun <T : Scalable<T>> T?.scale(factor: Double?): T? = this.apply(factor) { x, y -> x.scale(y) }

/**
 * Performs the given operation on the given inputs if they are both non-null.  If one input is null, the other input is
 * returned; if both are null, null is returned.
 *
 * @receiver some nullable object
 * @param other another nullable object
 * @param opFunction the operation to perform
 * @return the result as described
 * @param <T> the type of objects being operated upon
 */
fun <T> T?.op(other: T?, opFunction: BinaryOperator<T>): T? {
    return if (this == null) other
    else if (other == null) this
    else opFunction.apply(this, other)
}

/**
 * Applies the given transformation to an object if the other input is non-null.  If `this` is null, then null is
 * returned; if `other` is null, then `this` will be returned.
 *
 * @receiver some nullable object
 * @param other another nullable object
 * @param opFunction the operation to perform
 * @return the result as described
 * @param <T> the type of objects being operated upon
 */
fun <T, U> T?.apply(other: U?, opFunction: BiFunction<T, U, T>): T? {
    return if (this == null) null
    else if (other == null) this
    else opFunction.apply(this, other)
}

/**
 * Determines if both objects are null.
 *
 * @param a some nullable object
 * @param b another nullable object
 * @return true if both objects are null; false if either or both is non-null
 */
private fun allNull(a: Any?, b: Any?) = (a == null && b == null)
